File: app/src/main/java/com/friday/ar/ui/FileSelectorActivity.java
package com.friday.ar.ui;

import android.animation.ObjectAnimator;
import android.animation.StateListAnimator;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.friday.ar.FridayApplication;
import com.friday.ar.R;
import com.friday.ar.Theme;
import com.friday.ar.activities.FridayActivity;
import com.friday.ar.plugin.PluginVerticalListAdapter;
import com.friday.ar.util.FileUtil;
import com.google.android.material.appbar.AppBarLayout;
import com.sothree.slidinguppanel.SlidingUpPanelLayout;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.jar.JarFile;

import static android.Manifest.permission.WRITE_EXTERNAL_STORAGE;
import static android.content.pm.PackageManager.PERMISSION_GRANTED;
import static com.friday.ar.service.PluginIndexer.verify;

//TODO:Add support for selecting files
public class FileSelectorActivity extends FridayActivity {
    private static final String LOGTAG = "FileSelector";
    private RecyclerView fileList;
    private SlidingUpPanelLayout slidingUpPanelLayout;
    private Button openPlugin;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(Theme.getCurrentAppTheme(this));
        super.onCreate(savedInstanceState);
        requestPermissions(new String[]{WRITE_EXTERNAL_STORAGE}, 8001);
        supportRequestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.activity_file_selector);
        setSupportActionBar(findViewById(R.id.toolbar));
        getSupportActionBar().setTitle(R.string.open_plugin_file);
        slidingUpPanelLayout = findViewById(R.id.contentPane);
        TextView indexingStatus = findViewById(R.id.indexing_status);
        openPlugin = findViewById(R.id.select_file);
        setResult(RESULT_CANCELED);
        FridayApplication app = (FridayApplication) getApplication();
        indexingStatus.setText(app.getIndexedFiles().size() != 0 ?
                getResources().getQuantityString(R.plurals.pluginInstaller_indexingStatus, app.getIndexedFiles().size(), app.getIndexedFiles().size()) :
                getString(R.string.pluginInstaller_noItemsIndexed));
        slidingUpPanelLayout.setDragView(R.id.buttonbar);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == 8001 && Arrays.equals(grantResults, new int[]{PERMISSION_GRANTED})) {
            LinearLayoutManager fileListLayoutManager = new LinearLayoutManager(this, RecyclerView.VERTICAL, false);
            fileList = findViewById(R.id.fileList);
            fileList.setLayoutManager(fileListLayoutManager);
            fileList.setAdapter(new FileSystemArrayAdapter(Environment.getExternalStorageDirectory()));
            AppBarLayout appBarLayout = findViewById(R.id.appbar);
            StateListAnimator stateListAnimator = new StateListAnimator();
            fileList.setOnScrollChangeListener((v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
                if (fileListLayoutManager.findFirstCompletelyVisibleItemPosition() == 0) {
                    stateListAnimator.addState(new int[0], ObjectAnimator.ofFloat(appBarLayout, "elevation", 0));
                    appBarLayout.setStateListAnimator(stateListAnimator);
                } else {
                    StateListAnimator stateListAnimator1 = new StateListAnimator();
                    stateListAnimator1.addState(new int[0], ObjectAnimator.ofFloat(appBarLayout, "elevation", 16));
                    appBarLayout.setStateListAnimator(stateListAnimator1);
                }
            });
            RecyclerView indexedFilesList = findViewById(R.id.indexedFilesList);
            indexedFilesList.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
            indexedFilesList.setAdapter(new PluginVerticalListAdapter(this, ((FridayApplication) getApplication()).getIndexedFiles()));
        }
    }

    @Override
    public void onBackPressed() {
        FileSystemArrayAdapter fileListAdapter = ((FileSystemArrayAdapter) fileList.getAdapter());
        if (slidingUpPanelLayout.getPanelState() == SlidingUpPanelLayout.PanelState.EXPANDED) {
            slidingUpPanelLayout.setPanelState(SlidingUpPanelLayout.PanelState.COLLAPSED);
        } else if (fileListAdapter.sourceFile.getParentFile() != null && fileListAdapter.sourceFile.getParentFile().canRead()) {
            Log.d(LOGTAG, "parent:" + fileListAdapter.sourceFile.getParentFile().getPath());
            fileListAdapter.setDirectoryList(fileListAdapter.sourceFile.getParentFile());
        } else {
            super.onBackPressed();
        }
    }

    class FileSystemArrayAdapter extends RecyclerView.Adapter<FileViewHolder> {
        File sourceFile;
        List<File> directoryList;

        FileSystemArrayAdapter(File startFrom) {
            Log.d(LOGTAG, "starting file array adapter");
            setDirectoryList(startFrom);
        }

        @NonNull
        @Override
        public FileViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            return new FileViewHolder(LayoutInflater.from(getApplicationContext()).inflate(R.layout.file_selector_list_item, parent, false));
        }

        @SuppressLint("SetTextI18n")
        @Override
        public void onBindViewHolder(@NonNull FileViewHolder holder, int position) {
            Log.d(LOGTAG, "creating item " + position);
            File directoryFileItem = directoryList.get(position);
            holder.fileName.setText(directoryFileItem.getName());
            if (directoryFileItem.isDirectory() && !directoryFileItem.isFile()) {
                holder.fileIcon.setImageDrawable(getDrawable(R.drawable.ic_twotone_folder_24px));
                holder.fileSize.setVisibility(View.GONE);
                holder.root.setOnClickListener(v -> {
                    setResult(RESULT_CANCELED);
                    Log.d(LOGTAG, directoryFileItem.getPath());
                    if (directoryFileItem.canRead()) setDirectoryList(directoryFileItem);
                });
            } else {
                //TODO: Add visible sign that marks an file as selected
                holder.fileIcon.setImageDrawable(getDrawable(R.drawable.ic_twotone_insert_drive_file_24px));
                holder.root.setOnClickListener(v -> {
                    holder.fileIcon.animate().scaleX(0f).scaleY(0f).setDuration(200).setInterpolator(new AccelerateInterpolator()).start();
                    holder.iconProgress.animate().scaleX(1f).scaleY(1f).setDuration(200).setInterpolator(new DecelerateInterpolator()).start();
                    //Running verification on an file can take some time, so we run it on another thread
                    new Thread(() -> {
                        try {
                            if (FileUtil.getFileExtension(directoryFileItem).equals(".jar") && verify(new JarFile(directoryFileItem))) {
                                Intent resultIntent = new Intent();
                                resultIntent.setData(Uri.fromFile(directoryFileItem));
                                setResult(RESULT_OK, resultIntent);
                                openPlugin.setEnabled(true);
                                openPlugin.setOnClickListener((view) -> finish());
                            } else {
                                setResult(RESULT_CANCELED);
                                openPlugin.setEnabled(false);
                            }
                        } catch (IOException e) {
                            Log.e(LOGTAG, e.getMessage(), e);
                        }
                        runOnUiThread(() -> {
                            holder.fileIcon.animate().scaleX(1f).scaleY(1f).setDuration(200).setInterpolator(new DecelerateInterpolator()).start();
                            holder.iconProgress.animate().scaleX(0f).scaleY(0f).setDuration(200).setInterpolator(new AccelerateInterpolator()).start();
                        });
                    }).start();
                });
            }

        }

        @Override
        public int getItemCount() {
            return sourceFile.listFiles().length;
        }

        public void setDirectoryList(File directoryFile) {
            if (sourceFile != null) {
                notifyItemRangeRemoved(0, sourceFile.listFiles() != null ? sourceFile.listFiles().length : 0);
            }
            sourceFile = directoryFile;
            directoryList = Arrays.asList(directoryFile.listFiles());
            directoryList.sort((file1, file2) -> {
                if (file1.isDirectory() && file2.isFile())
                    return -1;
                if (file1.isDirectory() && file2.isDirectory()) {
                    return 0;
                }
                if (file1.isFile() && file2.isFile()) {
                    return 0;
                }
                return 1;
            });
            notifyItemRangeInserted(0, directoryList.size());
        }
    }

    class FileViewHolder extends RecyclerView.ViewHolder {
        LinearLayout root;
        ImageView fileIcon;
        ProgressBar iconProgress;
        TextView fileName;
        TextView fileSize;

        FileViewHolder(@NonNull View itemView) {
            super(itemView);
            root = itemView.findViewById(R.id.itemRootLayout);
            fileIcon = itemView.findViewById(R.id.fileIcon);
            iconProgress = itemView.findViewById(R.id.iconProgress);
            fileSize = itemView.findViewById(R.id.fileSize);
            fileName = itemView.findViewById(R.id.fileName);
        }
    }
}

File: app/src/main/java/com/friday/ar/ui/FileSelectorActivity.kt
package com.friday.ar.ui

import android.Manifest.permission.WRITE_EXTERNAL_STORAGE
import android.animation.ObjectAnimator
import android.animation.StateListAnimator
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager.PERMISSION_DENIED
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.animation.DecelerateInterpolator
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.friday.ar.FridayApplication
import com.friday.ar.R
import com.friday.ar.Theme
import com.friday.ar.activities.FridayActivity
import com.friday.ar.plugin.PluginVerticalListAdapter
import com.google.android.material.appbar.AppBarLayout
import com.sothree.slidinguppanel.SlidingUpPanelLayout
import kotlinx.android.synthetic.main.activity_file_selector.*
import java.io.File
import java.util.Arrays
import kotlin.Comparator

class FileSelectorActivity : FridayActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(Theme.getCurrentAppTheme(this))
        super.onCreate(savedInstanceState)
        requestPermissions(arrayOf(WRITE_EXTERNAL_STORAGE), 8001)
        supportRequestWindowFeature(Window.FEATURE_NO_TITLE)
        setContentView(R.layout.activity_file_selector)
        setSupportActionBar(toolbar)
        supportActionBar!!.setTitle(R.string.open_plugin_file)
        setResult(Activity.RESULT_CANCELED)
        val app = application as FridayApplication
        indexing_status.text = if (app.indexedInstallablePluginFiles.size != 0)
            resources.getQuantityString(R.plurals.pluginInstaller_indexingStatus, app.indexedInstallablePluginFiles.size, app.indexedInstallablePluginFiles.size)
        else
            getString(R.string.pluginInstaller_noItemsIndexed)
        contentPane.setDragView(R.id.buttonbar)
        select_file.setOnClickListener { finish() }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == 8001 && Arrays.equals(grantResults, intArrayOf(PERMISSION_GRANTED))) {
            val fileListLayoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
            fileList.layoutManager = fileListLayoutManager
            fileList.adapter = FileSystemArrayAdapter(Environment.getExternalStorageDirectory())
            val appBarLayout = findViewById<AppBarLayout>(R.id.appbar)
            val stateListAnimator = StateListAnimator()
            fileList.setOnScrollChangeListener { _, _, _, _, _ ->
                if (fileListLayoutManager.findFirstCompletelyVisibleItemPosition() == 0) {
                    stateListAnimator.addState(IntArray(0), ObjectAnimator.ofFloat(appBarLayout, "elevation", 0f))
                    appBarLayout.stateListAnimator = stateListAnimator
                } else {
                    val stateListAnimator1 = StateListAnimator()
                    stateListAnimator1.addState(IntArray(0), ObjectAnimator.ofFloat(appBarLayout, "elevation", 8f))
                    appBarLayout.stateListAnimator = stateListAnimator1
                }
            }
            val indexedFilesList = findViewById<RecyclerView>(R.id.indexedFilesList)
            indexedFilesList.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
            indexedFilesList.adapter = PluginVerticalListAdapter(this, (application as FridayApplication).indexedInstallablePluginFiles)
        } else if (Arrays.equals(grantResults, intArrayOf(PERMISSION_DENIED))) {
            fileListFlipper.displayedChild = 2
        }
    }

    override fun onBackPressed() {
        val fileListAdapter = fileList.adapter as FileSystemArrayAdapter?
        if (contentPane.panelState == SlidingUpPanelLayout.PanelState.EXPANDED) {
            contentPane.panelState = SlidingUpPanelLayout.PanelState.COLLAPSED
        } else if (fileListAdapter != null &&
                fileListAdapter.sourceFile!!.parentFile != null &&
                fileListAdapter.sourceFile!!.parentFile.canRead()) {
            fileListAdapter.setDirectoryList(fileListAdapter.sourceFile!!.parentFile)
        } else {
            super.onBackPressed()
        }
    }
    internal inner class FileSystemArrayAdapter(startFrom: File) : RecyclerView.Adapter<FileViewHolder>() {
        var sourceFile: File? = null
        private lateinit var directoryList: List<File>

        init {
            Log.d(LOGTAG, "starting file array adapter")
            setDirectoryList(startFrom)
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FileViewHolder {
            return FileViewHolder(LayoutInflater.from(applicationContext).inflate(R.layout.file_selector_list_item, parent, false))
        }

        @SuppressLint("SetTextI18n")
        override fun onBindViewHolder(holder: FileViewHolder, position: Int) {
            Log.d(LOGTAG, "creating item $position")
            val directoryFileItem = directoryList[position]
            holder.fileName.text = directoryFileItem.name
            if (directoryFileItem.isDirectory && !directoryFileItem.isFile) {
                holder.fileIcon.setImageDrawable(getDrawable(R.drawable.ic_twotone_folder_24px))
                holder.fileSize.visibility = View.GONE
                select_file.isEnabled = false
                holder.root.setOnClickListener {
                    setResult(Activity.RESULT_CANCELED)
                    Log.d(LOGTAG, directoryFileItem.path)
                    if (directoryFileItem.canRead()) setDirectoryList(directoryFileItem)
                }
            } else {
                //TODO: Add visible sign that marks an file as selected
                holder.fileIcon.setImageDrawable(getDrawable(R.drawable.ic_twotone_insert_drive_file_24px))
                holder.root.setOnClickListener {
                    select_file.isEnabled = true
                    holder.check.visibility = View.VISIBLE
                    holder.check.animate().scaleX(1f).scaleY(1f).setDuration(200).setInterpolator(DecelerateInterpolator()).start()
                    val resultIntent = Intent()
                    resultIntent.data = Uri.fromFile(directoryFileItem)
                    setResult(Activity.RESULT_OK, resultIntent)
                }
            }

        }

        override fun getItemCount(): Int {
            return sourceFile!!.listFiles().size
        }

        fun setDirectoryList(directoryFile: File) {
            if (sourceFile != null) {
                notifyItemRangeRemoved(0, if (sourceFile!!.listFiles() != null) sourceFile!!.listFiles().size else 0)
                if (sourceFile!!.listFiles() == null || sourceFile!!.listFiles().isEmpty()) {
                    fileListFlipper.displayedChild = 1
                } else {
                    fileListFlipper.displayedChild = 0
                }
            }
            sourceFile = directoryFile
            directoryList = Arrays.asList(*directoryFile.listFiles())
            directoryList.sortedWith(Comparator { file1: File, file2: File ->
                if (file1.isDirectory && file2.isFile)
                    return@Comparator -1
                if (file1.isDirectory && file2.isDirectory) {
                    return@Comparator 0
                }
                if (file1.isFile && file2.isFile) {
                    return@Comparator 0
                }
                1
            }
            )
            notifyItemRangeInserted(0, directoryList.size)
        }
    }

    internal inner class FileViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var root: LinearLayout = itemView.findViewById(R.id.itemRootLayout)
        var fileIcon: ImageView = itemView.findViewById(R.id.fileIcon)
        var check: ImageView = itemView.findViewById(R.id.check)
        var fileName: TextView = itemView.findViewById(R.id.fileName)
        var fileSize: TextView = itemView.findViewById(R.id.fileSize)

    }

    companion object {
        private const val LOGTAG = "FileSelector"
    }
}
